### PR TITLE
Proposed implementation

### PR DESCRIPTION
**Pull request by shigmas**
https://docs.google.com/document/d/1Zy6FGA-p7sur8aH-WZYx-DN_5Y2GgME4DjtFdAjYmWU/edit?usp=sharing

This is the reimplementation of the device-sdk. There are still a few pieces left, but the examples are running, so it's time to get some eyes on this.

o MQTT client rewritten. It requires more work for the user, and more responsibility has been pushed to the user as well.
o Rest code is basically untouched. 
o echo examples re-written, ddp examples removed. 
o removed session, kv_cache
o packaged renamed from mode to mode_client.

Items remaining:
o Update the rest code to match the style of the mqtt stuff
o Remove the auth stuff for the tests.
o Clean up the examples. It maybe useful to use some of the code as helpers, since we removed a lot of the nice stuff from session, but we don't want to reimplement it all the time.

### CODE REVIEW

File: core.go [hunk 1]
```diff
@@ -9,18 +9,15 @@ Both incoming commands and outgoing events are queued. If the connection is
 disrupted, commands already in the queue will be processed. Likewise,
 events already in the queue will be delivered when the connection resumes.
 */
-package mode
+package mode_client
```
Comment (takeshinoda) [inline core.go]: I like this changing name. `mode` is too big in namespace. But underscore isn't recommended in Go lang.

https://blog.golang.org/package-names

So........ for example `modesdk` ...? `gomode` ....? ummm

File: examples/echo/main.go [hunk 2]
```diff
@@ -5,64 +5,165 @@ command.
...
-	mode.SetCommandHandler("doEcho", doEcho)
+func main() {
+	var pingWg sync.WaitGroup
 
-	mode.SetDefaultCommandHandler(func(_ *mode.DeviceContext, cmd *mode.DeviceCommand) {
-		fmt.Printf("Received unknown command %s\n", cmd.Action)
-	})
+	dc := &mode_client.DeviceContext{
+		DeviceID:  6040,
+		AuthToken: "v1.ZHw2MDQw.1569621662.113e3dd1c0d5024816a41c0264fa0a734dfe31db218da0a480ceceee9961123348dc1aaf563480108b6f438c0710bbe244972f979e2ee08164a5d9c4c8a306d0633a96bff4edaec9",
```
Comment (takeshinoda) [inline examples/echo/main.go]: Btw, this token is available?

File: examples/echo/main.go [hunk 3]
```diff
@@ -5,64 +5,165 @@ command.
...
+		}
+
+		fmt.Println("Ping sent and ack'ed")
+		// Wait for the timer to before we loop again
+		<-ticker.C
 	}
+}
 
-	eventData := map[string]interface{}{"msg": params.Msg}
+func waitForAck(delegate *mode_client.ModeMqttDelegate) uint16 {
```
Comment (takeshinoda) [inline examples/echo/main.go]: `ModeMqttDelegate` method receivers are using a value. But this method is using a pointer. 
Could you tell me what is reason. Just coping cost?

File: examples/echo/main.go [hunk 4]
```diff
@@ -5,64 +5,165 @@ command.
...
 
-	if err := mode.SendEvent("echo", eventData, mode.QOSAtLeastOnce); err != nil {
-		fmt.Printf("Failed to send event: %s\n", err.Error())
+	// Block on the return channel or timeout
+	select {
+	case queueRes := <-delegate.QueueAckCh:
+		if queueRes.Err != nil {
+			fmt.Printf("Queued request failed: %s\n", queueRes.Err)
+		} else {
+			return queueRes.PacketId
```
Comment (takeshinoda) [inline examples/echo/main.go]: How does `mode_client.MqttClient` know sent `PacketId` of `client.PublishEvent(event)`? Or is there know to way sent `packetId`?

File: examples/echo/main.go [hunk 5]
```diff
@@ -5,64 +5,165 @@ command.
...
-	})
+	cmdQueue := make(chan *mode_client.DeviceCommand, 16)
+	kvSyncQueue := make(chan *mode_client.KeyValueSync, 16)
+	delegate := mode_client.NewModeMqttDelegate(dc, cmdQueue, kvSyncQueue)
 
-	if d, err := dc.GetInfo(); err == nil {
-		fmt.Printf("Running as device %v\n", d)
-	} else {
-		fmt.Printf("Failed to get device info: %s\n", err.Error())
+	client := mode_client.NewMqttClient(modeMqttHost, modeMqttPort, nil,
```
Comment (takeshinoda) [inline examples/echo/main.go]: Sorry, where is `NewMqttClient` define?

File: mqtt_mode.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+	// BulkData represents a batch of opaque data to be sent to the MODE cloud.
+	DeviceBulkData struct {
+		StreamID string
+		Blob     []byte
+		Qos      QOSLevel // not exported to serializer
+	}
+
+	// Implements the MqttDelegate
+	ModeMqttDelegate struct {
```
Comment (takeshinoda) [inline mqtt_mode.go:59]: `ModeMqttDelegate` can be seen from the side of MqttClient?
Normally it leaves MQTT communication to MqttClient instance?

File: mqtt.go [hunk 7]
```diff
@@ -19,598 +24,620 @@ const (
...
-		if ack.PacketID == p.PacketID {
-			logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
-			return nil
-		}
-
-		msg := fmt.Sprintf("sendEvent() received packet that does not match the Ack packet ID (%d) and the expected packet ID (%d).",
-			ack.PacketID, p.PacketID)
-		logInfo("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("event delivery unmatched Ack Packet: %s", msg))
+	case conn.outPktCh <- p:
```
Comment (takeshinoda) [inline mqtt.go:442]: If QoS level is one then client user should be wait PUBACK by self using delegator?

File: examples/echo/main.go (hunk 4 above)
Comment (takeshinoda) [inline examples/echo/main.go:89]: It there possible that `PacketID` of `PUBACK` order different from sending?

File: core.go (hunk 1 above)
Comment (shigmas) [reply core.go]: I see. It makes me think more about the structure. mode_client.MqttClient also looks strange. modeapi?

File: examples/echo/main.go (hunk 2 above)
Comment (shigmas) [reply examples/echo/main.go]: It is my device and key on staging. It's good for lazy testing and I will remove it before merging.

File: examples/echo/main.go (hunk 3 above)
Comment (shigmas) [reply examples/echo/main.go]: Yes. In my habit, I would have used "const &" from C++. In this case, I should just pass the channel.

File: mqtt_mode.go (hunk 6 above)
Comment (shigmas) [reply mqtt_mode.go:59]: MqttClient will call methods on MqttDelegate, but it doesn't know about ModeMqttDelegate. Is that your question?

File: mqtt.go (hunk 7 above)
Comment (shigmas) [reply mqtt.go:442]: The client will wait for the PUBACK by the packet ID. The client should wait for the PUBACK.

File: examples/echo/main.go (hunk 4 above)
Comment (shigmas) [reply examples/echo/main.go:89]: Yes, it's very possible. So, if the client has sent many requests, it will get several acks from the delegate's QueueAckCh channel. The client will have to store the packet ID's and match them to know which requests have been ACK'ed.

File: core.go (hunk 1 above)
Comment (ethankan) [reply core.go]: How about....
- modedev
- modecli
- mdevice
- mordecai

File: examples/echo/main.go [hunk 8]
```diff
@@ -5,64 +5,165 @@ command.
...
 	"github.com/moderepo/device-sdk-go/v2"
 )
 
-func doEcho(_ *mode.DeviceContext, cmd *mode.DeviceCommand) {
-	// Command parameters are JSON encoded and can be retrieved as follows.
-	var params struct {
-		Msg string `json:"msg"`
-	}
+var (
+	// My cloudmqtt.com server)
```
Comment (takjn) [inline examples/echo/main.go]: Please fix the comment.

File: examples/echo/main.go [hunk 9]
```diff
@@ -5,64 +5,165 @@ command.
...
+	modeMqttHost = "staging-api.corp.tinkermode.com"
+	modeMqttPort = 8883
+	modeUseTLS   = true
+)
+
+func pingLoop(client *mode_client.MqttClient, timeout time.Duration,
+	pingRecv <-chan bool, gotDisconnected chan<- bool, doDisconnect <-chan bool,
+	wg sync.WaitGroup) {
+
+	wg.Add(1)
```
Comment (takjn) [inline examples/echo/main.go]: The main goroutine should call Add function.

File: examples/echo/main.go (hunk 2 above)
Comment (takjn) [reply examples/echo/main.go]: Here is a public repository. Please pay attention to your credentials.

File: mqtt_test.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+)
+
+var (
+	// My cloudmqtt.com server)
+	testMqttHost   = "tailor.cloudmqtt.com"
```
Comment (takjn) [inline mqtt_test.go]: Please change the host.

File: mqtt_test.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+	subRecvCh     chan MqttSubData
+	queueAckCh    chan MqttQueueResult
+	pingAckCh     chan bool
+}
+
+func newTestMqttDelegate() TestMqttDelegate {
+	return TestMqttDelegate{
+		// This is a user level password, which I can change, but it still
+		// shouldn't be in here. But, how else do you do automated tests in a
+		// public server?
```
Comment (takjn) [inline mqtt_test.go:25]: How about launching a fake server instead of using a public server?
e.g.
https://github.com/moderepo/device-sdk-go/blob/53c2e26d2a5677f5e745399d71ba64968a4f32b7/session_test.go#L64

File: core.go [hunk 12]
```diff
@@ -29,50 +26,12 @@ import (
 const (
 	sdkVersion      = "1.0"
```
Comment (takjn) [inline core.go]: Can you increase the version?

File: core.go [hunk 13]
```diff
@@ -9,18 +9,15 @@ Both incoming commands and outgoing events are queued. If the connection is
 disrupted, commands already in the queue will be processed. Likewise,
 events already in the queue will be delivered when the connection resumes.
```
Comment (takjn) [inline core.go]: Can you update the package comment?

File: core.go [hunk 14]
```diff
@@ -147,54 +65,12 @@ type (
...
-
-	// BulkData represents a batch of opaque data to be sent to the MODE cloud.
-	DeviceBulkData struct {
-		StreamID string
-		Blob     []byte
-		qos      QOSLevel // not exported to serializer
-	}
-
 	// BulkData represents a batch of opaque data to be sent to the MODE cloud.
 	DeviceSyncedBulkData struct {
```
Comment (takjn) [inline core.go]: I could not find other reference in this SDK. Should we remove this struct?

File: core.go [hunk 15]
```diff
@@ -147,54 +65,12 @@ type (
 		DeviceClass string `json:"deviceClass"`
 	}
 
-	// DeviceCommand represents a command received from the MODE cloud.
```
Comment (takjn) [inline core.go:150]: IMO, Command, Event and Key-value is the first-class object in MODE device. And we can also access these objects through the REST API in addition to MQTT. Staying here may be better than moving.

File: core.go (hunk 1 above)
Comment (takjn) [reply core.go]: I think we don't need to change the package name for now. Because this package is in `device-go-sdk`. `mode` is not too big in the context.

File: examples/echo/main.go [hunk 16]
```diff
@@ -5,64 +5,165 @@ command.
...
-	}
+var (
+	// My cloudmqtt.com server)
+	modeMqttHost = "staging-api.corp.tinkermode.com"
+	modeMqttPort = 8883
+	modeUseTLS   = true
+)
+
+func pingLoop(client *mode_client.MqttClient, timeout time.Duration,
+	pingRecv <-chan bool, gotDisconnected chan<- bool, doDisconnect <-chan bool,
```
Comment (takjn) [inline examples/echo/main.go]: Can you replace `doDisconnect` with receiving the parent's context?
I think using `context.Context` is more common method to stop goroutines.

**Review by takjn — COMMENTED** [review]
To be honest, I haven't finished reading the entire code step by step. As far as I read the example code, I feel the API may be difficult for application developers to use.

Also, this may be off topic, this makes me think more about the SDK. In my understanding, this is an SDK for device. We might need to provide another SDK in the future. (e.g. SDK for application server, not for device, using REST API, not MQTT).

File: mqtt_mode.go (hunk 6 above)
Comment (takeshinoda) [reply mqtt_mode.go:59]: Yes! Thank you!

**Comment by takeshinoda** [comment]
Thank you Masa-san! I could understand your proposals (Maybe)!
Basically, I like this implementation, because this design is back pressure less design, I think.

I have some opinions.

1. More better that it exists layer of handling `PUBACK` for this SDK user.
2. This SDK needs a Keeping connection feature. Current SDK implementation has keep connection feature.
3. Perhaps it is good if it separate 2 packages. 
  First, low layer SDK. It is only device authentication and exposure transmit sequence. Maybe same this PR. `github.com/moderepo/device-sdk-go/v3/mqtt`. Second, high layer SDK. It is support helper methods of synchronous call (== QoS1) and  DDP and Command... Maybe same current (v2) SDK.  `github.com/moderepo/device-sdk-go/v3`

File: examples/echo/main.go (hunk 2 above)
Comment (shigmas) [reply examples/echo/main.go]: I also think this discussion is strange to have on a public repository, so I will check with Ethan and fork this into a private repository. And, of course, I'll invalidate the keys after I'm done with them.

**Comment by shigmas** [comment]
@takeshinoda @takjn Thanks for your feedback. My goal was to decrease the "magic". But I could see in the example (and the tests) that there is too much boilerplate code that has to be written.

Takeshi - for point 3, that is exactly my plan! But the high level SDK will live in applications, so this is a good discussion to see which code should remain here, and what code I will move to our applications repository. Right now the lower level code is too hard to use.

File: core.go (hunk 15 above)
Comment (shigmas) [reply core.go:150]: I have moved them into mqtt_mode.go, so it isn't totally gone. But, your point is good about key-value's. I will improve the handling in this library. But, what do you think about moving the cache into a common applications library?

File: examples/echo/main.go (hunk 9 above)
Comment (shigmas) [reply examples/echo/main.go]: Oh, I see. It prevents race conditions in the general case. Thanks.

File: mqtt_test.go (hunk 11 above)
Comment (shigmas) [reply mqtt_test.go:25]: Ah. I should have looked at the old code more closely. Okay - I've made a new fake server which is able to do more test cases.

File: core.go (hunk 12 above)
Comment (shigmas) [reply core.go]: Sure! Looks like we even skipped 2!

File: core.go (hunk 13 above)
Comment (shigmas) [reply core.go]: I have removed it, and left TBD, since no comments are better than wrong comments.

File: core.go (hunk 14 above)
Comment (shigmas) [reply core.go]: Yes, at least for now.

File: examples/echo/main.go [hunk 17]
```diff
@@ -5,64 +5,165 @@ command.
...
+
+	for {
+		if err := client.Ping(); err != nil {
+			gotDisconnected <- true
+			return
+		}
+
+		// Wait for the ack, or timeout
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
```
Comment (ecnerwalawrence) [inline examples/echo/main.go:43]: this should be outside of the go routine.  Please do @takjn  suggestion with context.

File: core.go (hunk 15 above)
Comment (takjn) [reply core.go:150]: I agree about moving the cache into a common applications library, in other word, a high level SDK.
I think the high level SDK also should be public. Because many developer will want to use the feature.

File: core.go [hunk 18]
```diff
@@ -320,14 +184,6 @@ func (kv *KeyValue) String() string {
...
-// A special JSON decoder that makes sure numbers in command parameters
-// are preserved (avoid turning integers into floats).
-func decodeOpaqueJSON(b []byte, v interface{}) error {
-	decoder := json.NewDecoder(bytes.NewReader(b))
-	decoder.UseNumber()
-	return decoder.Decode(v)
-}
-
 // BindParameters maps the command parameters from JSON to the provided struct.
 func (cmd *DeviceCommand) BindParameters(v interface{}) error {
```
Comment (takjn) [inline core.go:188]: Can you move methods on `DeviceCommand` struct and  `KeyValue` struct to `mqtt_mode.go`?

File: mqtt.go [hunk 19]
```diff
@@ -19,598 +24,613 @@ const (
...
+	// - data needed to establish a connection
+	// - Callback method to create data to publish
+	// - channels to communicate data back to a receiver.
+	// These methods should be const, since we might call them frequently.
+	// I think this could be broken up - maybe a struct for the channels, with
+	// another interface (and the struct implement the interface if desired).
+	// But this is sufficient for now.
+	MqttDelegate interface {
+		// Returns authentication information
+		AuthInfo() (username string, password string)
```
Comment (takjn) [inline mqtt.go:67]: I feel unconformable that MqttDelegate interface handles AuthInfo().
IMO MqttClient struct should hold username and password.

File: mqtt.go [hunk 20]
```diff
@@ -19,598 +24,613 @@ const (
...
 
-func (mc *mqttConn) sendPacket(p packet.Packet) error {
-	if err := mc.stream.Write(p); err != nil {
-		logError("[MQTT] failed to send %s packet: %s", p.Type(), err.Error())
-		return err
+// Creates a client and connects. A client is invalid if not connected, and
+// you need to create a new client to reconnect and subscribe
+// a pingInterval of zero means no keepalive pings
+func NewMqttClient(mqttHost string, mqttPort int, tlsConfig *tls.Config,
+	useTLS bool, delegate MqttDelegate) *MqttClient {
```
Comment (takjn) [inline mqtt.go:145]: Here is why I'm not comfortable MqttDelegate manage username and password. My understanding is that TLS is part of the MQTT authentication mechanism.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349852

My suggestion may be easier to understand for who familiar with MQTT.
```suggestion
	useTLS bool, username string, password string, delegate MqttDelegate) *MqttClient {
```

File: mqtt.go [hunk 21]
```diff
@@ -19,598 +24,613 @@ const (
...
+	}()
+
+	// Creates a new context, using the client's timeout
+	p := packet.NewDisconnectPacket()
+	_, err := client.conn.SendPacket(p)
+	if err != nil {
+		logError("[MQTT] failed to send packet: %s", err.Error())
+		return err
+	}
+	client.conn.Close()
```
Comment (takjn) [inline mqtt.go:234]: This line confused me at first.
Connect() method doesn't touch mqttConn. It just send a CONNECT packet.
Disconnect() method closes mqttConn in addition to sending a DISCONNECT packet.

However now I understand this is a good idea. So it will be better to connect the server in Connect() method in the same manner. We will be able to reconnect the same server easily.

File: mqtt.go [hunk 22]
```diff
@@ -19,598 +24,613 @@ const (
...
-	p.Message = packet.Message{
-		Topic:   fmt.Sprintf("/devices/%d/kv", mc.dc.DeviceID),
-		QOS:     qos,
-		Payload: payload,
+func (conn *mqttConn) writePacket(p packet.Packet) error {
+	// Since we have both queued writes, and blocking, we lock at the
+	// lowest level
+	conn.mutex.Lock()
+	defer conn.mutex.Unlock()
+	if err := conn.stream.Write(p); err != nil {
```
Comment (takjn) [inline mqtt.go:553]: Can you set a timeout?
I think we can set deadline to the connection.
https://golang.org/pkg/net/#IPConn.SetDeadline

File: mqtt.go [hunk 23]
```diff
@@ -19,598 +24,613 @@ const (
...
+		return err
+	}
+	client.conn.Close()
+
 	return nil
 }
 
-func (mc *mqttConn) subscribe(subs []mqttSubscription) error {
+// Add the subscriptions from the delegate
+func (client *MqttClient) Subscribe() error {
```
Comment (takjn) [inline mqtt.go:240]: We can not change the subscriptions with this implementation.
It may not be a problem for now but it will be better if we can change the subscription.

File: mqtt.go [hunk 24]
```diff
@@ -19,598 +24,613 @@ const (
...
+		// Disconnect is a special case. Server will close, so no response, but
+		// it's a packet type we expect, so don't let it fall to default
+		return nil
+	case packet.CONNACK:
+		return conn.connRespCh
+	case packet.SUBSCRIBE:
+		// Subscribes are blocking, but have packet ID's, and, theoretically,
+		// multiples can be sent, so we enable this at this level.
+		subPkt := p.(*packet.SubscribePacket)
+		subPkt.PacketID = conn.getPacketID()
```
Comment (takjn) [inline mqtt.go:604]: As you know SUBSCRIBE packet must contain a PacketID.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349761
I think this should be set in Subscribe method for readability.

File: mqtt.go [hunk 25]
```diff
@@ -19,598 +24,613 @@ const (
...
-	mc.stopPublisher = make(chan bool)
-	mc.err = make(chan error, 1)
-	mc.outPacket = make(chan packet.Packet, 1)
-	mc.puback = make(chan *packet.PubackPacket, 10)     // make sure this won't block
-	mc.pingresp = make(chan *packet.PingrespPacket, 10) // make sure this won't block
-
-	go mc.runPacketReader()
-	go mc.runPacketWriter()
-	go mc.runPublisher()
-	go mc.runPinger()
```
Comment (takjn) [inline mqtt.go:613]: Why did you remove pinger loop from `mqtt.go`?
PING is very specific to MQTT and all applications need to ping. It will be better if we can implement in `mqtt.go`.

File: mqtt.go [hunk 26]
```diff
@@ -19,598 +24,613 @@ const (
...
-
-	mc.outPacket <- p
-
-	logInfo("[MQTT] synced bulk data delivery packet ID %d", p.PacketID)
-	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
+func (conn *mqttConn) runPacketReader(wg *sync.WaitGroup) {
+	defer func() {
+		logInfo("[MQTT] packet reader is exiting")
+		wg.Done()
+		// Close the rest of our channels?
```
Comment (takjn) [inline mqtt.go:496]: Yes, we should.
How about using errgroup with context?
https://godoc.org/golang.org/x/sync/errgroup

**Review by takjn — COMMENTED** [review]
Thank you for the improvement. I added some comments.
- `core.go` looks good to me. Only one tiny suggestion.
- `mqtt.go` needs more discussion.
- `mqtt_mode.go` not reviewed yet. (sorry...)

File: core.go (hunk 15 above)
Comment (shigmas) [reply core.go:150]: In my opinion, that's fine. There is no secret algorithms and only our customers are using this code. I'll discuss with Ethan.

File: core.go (hunk 18 above)
Comment (shigmas) [reply core.go:188]: Yes, that's the better place.

File: mqtt.go (hunk 19 above)
Comment (shigmas) [reply mqtt.go:67]: Since we use the device ID in the topic strings, I put user and password into the delegate. But, that is our implementation detail. We can just pass the parameters in Connect().

File: core.go (hunk 15 above)
Comment (takjn) [reply core.go:150]: I see. That's fine.

File: mqtt.go (hunk 20 above)
Comment (shigmas) [reply mqtt.go:145]: I haven't thought much about TLS, so it's a good time to talk about it now. Your API change is good, but the TLS config is in the delegate, since it's not part of the device. What do you think, @takjn

File: mqtt.go (hunk 23 above)
Comment (shigmas) [reply mqtt.go:240]: For Mode use, we only have fixed topics, but it is not a general solution. I'll think about it more.

File: mqtt.go (hunk 22 above)
Comment (shigmas) [reply mqtt.go:553]: Oh, no timeout with a lock is quite bad. Thanks!

File: mqtt.go (hunk 24 above)
Comment (shigmas) [reply mqtt.go:604]: @takjn Publish packet ID's aren't set in Publish. It should be at least in the same struct. I don't mind moving all the code for packet ID's from the connection into the client, so Publish and Subscribe both set the packet ID.

Actually, after talking about guaranteed delivery, I think creating the packet ID might be better in the session code

File: mqtt.go (hunk 25 above)
Comment (shigmas) [reply mqtt.go:613]: I had written it, but Ethan had suggested to move those things into the library in applications. But, everyone would have to implement it. Ethan's suggestion was to only ping if we are idle, so maybe a goroutine to check recent activity and ping if haven't been busy lately.

File: mqtt.go (hunk 26 above)
Comment (shigmas) [reply mqtt.go:496]: I am closing them in another place now. But, I'll look at the errgroup to see if that's a better solution.

File: mqtt.go (hunk 19 above)
Comment (takjn) [reply mqtt.go:67]: I understand. Thank you.

File: mqtt.go (hunk 20 above)
Comment (takjn) [reply mqtt.go:145]: Thank you, @shigmas. I understand. How about `AuthInfo()` returns TLS config in addition to username and password?

File: mqtt.go (hunk 19 above)
Comment (takjn) [reply mqtt.go:67]: I understand. Thank you.

File: mqtt.go (hunk 24 above)
Comment (takjn) [reply mqtt.go:604]: > Actually, after talking about guaranteed delivery, I think creating the packet ID might be better in the session code

I agree the idea. Packet ID should not be in black-box for the publisher.

**Comment by takjn** [comment]
Awesome work and thank you for explanation. I understand.
I think it's time to start testing testing refactored device SDK.

**Comment by shigmas** [comment]
I'll also write error types so the client can easily respond to the error by the type
